### PR TITLE
fix: disallow leading zeros in RLP decoding

### PR DESCRIPTION
**Pull request by Rjected**
## Motivation

Currently `Uint` RLP decoding ignores leading zeros.

## Solution

Disallow leading zeros.

## PR Checklist

- [x] Added Tests
- [ ] Added Documentation
- [x] Updated the changelog

### CODE REVIEW

**Comment by prestwich** [comment]
This would be a breaking behavior change. Is there a motivating reason?

**Comment by Rjected** [comment]
> This would be a breaking behavior change. Is there a motivating reason?

Sorry, should have been more specific in the PR description. The main motivation is that leading zeros on RLP uints is defined to be invalid, and encoding would also never output uints encoded this way. This also makes writing certain kinds of fuzz tests possible, specifically fuzz tests that:
* Take fuzz data as input
* Decode TX or other struct with an int
* if successful, encode struct
* compare encoding output to bytes input

If we accept invalid RLP input, and are more permissive in decoding than encoding as a result, these fuzz tests quickly fail and do not act as meaningful fuzz tests.

**Comment by Rjected** [comment]
@prestwich just bumping, any thoughts on this given the context?

**Comment by prestwich** [comment]
The ethereum.org spec is clear as mud, but the yellowpaper appears to be unambiguous about this ([see (197)](https://ethereum.github.io/yellowpaper/paper.pdf#page=20)) . Accepting the non-minimal uint is off-spec, so we should fix it. However, because this silently changes behavior in a way that breaks downstream users, we should release it as a major version bump

**Review by prestwich — CHANGES_REQUESTED** [review]
Update the changelog :)

File: src/support/alloy_rlp.rs
Comment (recmo) [inline src/support/alloy_rlp.rs:1]: Link to spec on line 16 and 75 is outdated. Should be <https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp/>. Please fix.

File: src/support/alloy_rlp.rs [hunk 1]
```diff
@@ -77,6 +77,13 @@ impl<const BITS: usize, const LIMBS: usize> Decodable for Uint<BITS, LIMBS> {
     #[inline]
     fn decode(buf: &mut &[u8]) -> Result<Self, Error> {
         let bytes = Header::decode_bytes(buf, false)?;
+
+        // We only need to check if the first byte is zero to make sure there are no
+        // leading zeros
```
Comment (recmo) [inline src/support/alloy_rlp.rs:82]: Comment should mention that this is according to the RLP spec.

File: src/support/alloy_rlp.rs [hunk 2]
```diff
@@ -77,6 +77,13 @@ impl<const BITS: usize, const LIMBS: usize> Decodable for Uint<BITS, LIMBS> {
...
+        if !bytes.is_empty() && bytes[0] == 0 {
+            return Err(Error::LeadingZero);
+        }
```
Comment (recmo) [inline src/support/alloy_rlp.rs:85]: RLP spec explicitly says

> integers must be represented in big-endian binary form with no leading zeroes (thus making the integer value zero equivalent to the empty byte array).

meaning that the old accepting behaviour was not according to spec and a bug. Fixing bugs is not a breaking change.

There's 'robustness principle' argument that could be made to make this error optionally a warning (i.e. introduce a non-strict mode). But that seems more effort than it's worth. Unless someone volunteers the strict mode should be default.

**Review by recmo — APPROVED** [review]
LGTM. Bug fix. Not a breaking change.

**Comment by recmo** [comment]
> If we accept invalid RLP input, and are more permissive in decoding than encoding as a result, these fuzz tests quickly fail and do not act as meaningful fuzz tests.

This proposed test assumes that encoding is canonical. This assumption is only true when leading zeros are rejected.

> [ethereum.org] and [yellow paper]

Both yellow paper and ethereum.org spec clearly state leading zeros are an error. While neither mention it explicitly, the implicit goal is to make it such that there is only one canonical RLP encoding. This is important since it RLP encoding is often used to hash data structures. Having multiple valid hashes for the same data structure leads to all forms of madness.

**Comment by recmo** [comment]
> However, because this silently changes behavior in a way that breaks downstream users, we should release it as a major version bump

The accepting behavior was a bug, any downstream users depending on this bug are in error (as is clear by the RLP spec). IMO this a bug fix and not a breaking change.

**Comment by prestwich** [comment]
I've also submitted a PR to the RLP spec doc on the Ethereum website to disambiguate this

the worst part of this is how it breaks the layer boundary between RLP and "higher-order" protocols. You cannot accurately decode RLP without knowing whether the higher-order protocol contains positive integers

https://github.com/ethereum/ethereum-org-website/pull/11532

File: src/support/alloy_rlp.rs
Comment (Rjected) [reply src/support/alloy_rlp.rs:1]: fixed

File: src/support/alloy_rlp.rs (hunk 1 above)
Comment (Rjected) [reply src/support/alloy_rlp.rs]: quoted directly

**Review by DaniPopes — CHANGES_REQUESTED** [review]
Can you also update fastrlp? It should be an easy copy-paste

**Comment by Rjected** [comment]
> Can you also update fastrlp? It should be an easy copy-paste

good catch, will do

**Review by prestwich — CHANGES_REQUESTED** [review]
one more changelog nit

File: CHANGELOG.md [hunk 3]
```diff
@@ -32,6 +32,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - `leading_ones` failed for non-aligned sizes.
+- Restricted RLP decoding to match the RLP spec and disallow leading zeros ([#335])
```
[the remark this answers is not in the document]
Comment (Rjected) [reply CHANGELOG.md]: fixed!

**Review by prestwich — APPROVED** [review]
🙌

**Pull request merged**